jetty-server/src/main/java/org/eclipse/jetty/server/MultiPartFormInputStream.java

```diff
@@ -468,13 +476,24 @@ protected void throwIfError() throws IOException
      */
     protected void parse()
     {
-        // have we already parsed the input?
-        if (_parsed)
-            return;
-        _parsed = true;
+        synchronized (this)
+        {
+            switch (state)
+            {
+                case UNPARSED:
+                    state = State.PARSING;
+                    break;
+
+                case PARSED:
+                    return;
+
+                default:
+                    _err = new IllegalStateException(state.name());
```

Comment:
The spec is clear in the circumstances where IllegalStateException is allowed, and this case isn't one of them - you'll have to consider using IOException instead.

Also, in the case where there isn't a multipart/form-data content type, then if the same thread called getParts() twice, on the first call the state would remain as PARSING and return no parts, but the second call would result in an IllegalStateException - the result of calling the same method twice should be idempotent.

Reply:
What should `MultiPartFormInputStream.getParts()` do for a content type which is not multipart/form-data, should it return an empty list of parts?
I feel like that should be an IllegalArgumentException in the constructor or something.

Reply:
But the same thread calling `getParts()` twice with a non multipart/form-data content type, it would return from 
```java
// if its not a multipart request, don't parse it
if (_contentType == null || !_contentType.startsWith("multipart/form-data"))
    return;
```
end up in the finally block and move from PARSING->PARSED, so the second call to parse() would be a NOOP.